pkg/session/session.go

```diff
@@ -0,0 +1,165 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package session
+
+import (
+	"crypto"
+	"errors"
+	"github.com/ForgeRock/iot-edge/internal/client"
+	"github.com/ForgeRock/iot-edge/pkg/callback"
+	"net/url"
+	"time"
+)
+
+// Session holds session data
+type Session interface {
+
+	// Token returns the session token
+	Token() string
+
+	// HasRestrictedToken returns true if the session has a restricted token
+	HasRestrictedToken() bool
+
+	// SigningKey returns the signing key associated with a restricted SSO token
+	SigningKey() crypto.Signer
+
+	// Nonce returns the session nonce
+	Nonce() int
+
+	// IncrementNonce increments the session nonce
+	IncrementNonce()
+
+	// Valid returns true if the session is valid
+	Valid() (bool, error)
+
+	// Logout the session
+	Logout() error
+}
+
+type defaultSession struct {
+	connection client.Connection
+	token      string
+	nonce      int
+	key        crypto.Signer
+}
+
+func (s *defaultSession) Token() string {
+	return s.token
+}
+
+func (s *defaultSession) HasRestrictedToken() bool {
+	return s.key != nil
+}
+
+func (s *defaultSession) SigningKey() crypto.Signer {
+	return s.key
+}
+
+func (s *defaultSession) Nonce() int {
+	return s.nonce
+}
+
+func (s *defaultSession) IncrementNonce() {
+	s.nonce++
+}
+
+func (s *defaultSession) Valid() (bool, error) {
+	return s.connection.ValidateSession(s.token)
+}
+
+func (s *defaultSession) Logout() error {
+	return s.connection.LogoutSession(s.token)
+}
+
+type Builder struct {
+	url        *url.URL
+	realm      string
+	tree       string
+	timeout    time.Duration
+	connection client.Connection
+	handlers   []callback.Handler
+}
+
+func New() *Builder {
+	return &Builder{}
+}
+
+func (b *Builder) AuthenticateWith(handlers ...callback.Handler) *Builder {
+	b.handlers = handlers
+	return b
+}
+
+func (b *Builder) ConnectTo(url *url.URL) *Builder {
+	b.url = url
+	return b
+}
+
+func (b *Builder) WithConnection(connection client.Connection) *Builder {
+	b.connection = connection
+	return b
+}
+
+func (b *Builder) InRealm(realm string) *Builder {
+	b.realm = realm
+	return b
+}
+
+func (b *Builder) WithTree(tree string) *Builder {
+	b.tree = tree
+	return b
+}
+
+func (b *Builder) TimeoutRequestAfter(d time.Duration) *Builder {
+	b.timeout = d
+	return b
+}
+
+func (b *Builder) Create() (Session, error) {
+	var err error
+	if b.connection == nil {
+		if b.url == nil {
+			return nil, errors.New("url must be provided")
+		}
+		// TODO Hand in thing signer for CoAP connection self signed certificate
```

Comment:
It looks like we should be able to pass the thing's signer into the connection so that we don't have to generate a private key for the CoAP client. Is this correct? Do we have a story for this or should we add one?

Reply:
In fact, you had to supply a signer in the old `NewIECClient` method but this was lost when we moved over to the builder pattern. I added IOTEDGE-997 to make a note of this. 

I would like to avoid a CoAP specific builder function and have something that makes sense for both types of client if we can.